lib/util/pthread_create_blocking.c

```diff
@@ -0,0 +1,162 @@
+
+#include "pthread_create_blocking.h"
+
+struct wrapped_cookie {
+	/* User-supplied. */
+	void * (*start_routine)(void *);
+	void * arg;
+
+	/* Synchronization. */
+	pthread_mutex_t mutex;
+	pthread_cond_t cond;
+	int initialized;	/* 1 if initialized. */
+};
+
+/* Set the "initialized" flag and signal the condition variable. */
+static int
+set_initialized(struct wrapped_cookie * U)
+{
+	int rc;
+
+	/* Lock mutex. */
+	if ((rc = pthread_mutex_lock(&U->mutex)))
+		goto err0;
+
+	/* Set the "initialized" flag and signal the condition variable. */
+	U->initialized = 1;
+	if ((rc = pthread_cond_signal(&U->cond)))
+		goto err1;
+
+	/* Unlock mutex. */
+	if ((rc = pthread_mutex_unlock(&U->mutex)))
+		goto err0;
+
+	/* Success! */
+	return (0);
+
+err1:
+	/*
+	 * Don't record any errors in this clean-up; the existing value of rc
+	 * is more important.
+	 */
+	pthread_mutex_unlock(&U->mutex);
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Routine which is executed by pthread_create(). */
+static void *
+wrapped_thread(void * cookie)
+{
+	struct wrapped_cookie * U = cookie;
+	void * rc_voidp;
+
+	/* Copy the user-supplied parameters, because U will not be valid. */
+	void * (*start_routine)(void *) = U->start_routine;
+	void * arg = U->arg;
+
+	/* Indicate that the main thread can continue (which will free U). */
+	if (set_initialized(U))
+		return (NULL);
+
+	/* Run the desired routine. */
+	rc_voidp = start_routine(arg);
+
+	/* Done. */
+	return (rc_voidp);
+}
+
+/**
+ * pthread_create_blocking(thread, attr, start_routine, arg):
+ * Run pthread_create() and block until the the ${thread} has started.  The
+ * thread will execute ${start_routine} with ${arg} as its sole argument.
+ * When ${start_routine} finishes, make its returned value available via
+ * pthread_join().  If successful, return 0; otherwise return the error number.
+ */
+int
+pthread_create_blocking(pthread_t * restrict thread,
+    const pthread_attr_t * restrict attr,
+    void * (*start_routine)(void *), void * restrict arg)
+{
+	struct wrapped_cookie * U;
+	int rc;
+
+	/* Allocate our cookie and store parameters. */
+	if ((U = malloc(sizeof(struct wrapped_cookie))) == NULL) {
+		rc = errno;
+		goto err0;
+	}
+	U->start_routine = start_routine;
+	U->arg = arg;
+
+	/* Initialize & assign thread startup-related variables. */
+	if ((rc = pthread_mutex_init(&U->mutex, NULL)) != 0)
+		goto err1;
+	if ((rc = pthread_cond_init(&U->cond, NULL)) != 0)
+		goto err2;
+	U->initialized = 0;
+
+	/* Lock mutex. */
+	if ((rc = pthread_mutex_lock(&U->mutex)) != 0)
+		goto err2;
+
+	/* Create the thread. */
+	if ((rc = pthread_create(thread, attr, wrapped_thread, U)) != 0) {
+		/*
+		 * Don't record any pthread_mutex_unlock() error here; the
+		 * error from pthread_create() is more important.
+		 */
+		pthread_mutex_unlock(&U->mutex);
+		goto err3;
+	}
+
+	/* Wait for the thread to have started, then unlock the mutex. */
+	while (!U->initialized) {
+		/* Wait up to 1 second. */
```

Comment:
Oops, forgot to update that comment.

I'll hold off on changing it, pending your response about reverting 8f9e8ed (i.e. keeping the "check if pthreads mutex & condition variables are broken").